src/transformers/models/phi3/modeling_phi3.py

```diff
@@ -88,25 +89,35 @@ def _prepare_4d_causal_attention_mask_with_cache_position(
             Indices depicting the position of the input sequence tokens in the sequence.
         batch_size (`torch.Tensor`):
             Batch size.
+        config (`Phi3Config`):
+            The model's configuration class
+        past_key_values (`Cache`):
+            The cache class that is being used currently to generate
     """
     if attention_mask is not None and attention_mask.dim() == 4:
         # In this case we assume that the mask comes already in inverted form and requires no inversion or slicing.
         causal_mask = attention_mask
     else:
         causal_mask = torch.full((sequence_length, target_length), fill_value=min_dtype, dtype=dtype, device=device)
-        if sequence_length != 1:
-            causal_mask = torch.triu(causal_mask, diagonal=1)
-        causal_mask *= torch.arange(target_length, device=device) > cache_position.reshape(-1, 1)
+        exclude_mask = torch.arange(target_length, device=device) > cache_position.reshape(-1, 1)
+        if config.sliding_window is not None:
+            if not isinstance(past_key_values, SlidingWindowCache) or sequence_length > config.sliding_window:
+                exclude_mask.bitwise_or_(
+                    torch.arange(target_length, device=device)
+                    <= (cache_position.reshape(-1, 1) - config.sliding_window)
+                )
+        causal_mask *= exclude_mask
         causal_mask = causal_mask[None, None, :, :].expand(batch_size, 1, -1, -1)
         if attention_mask is not None:
             causal_mask = causal_mask.clone()  # copy to contiguous memory for in-place edit
+            if attention_mask.shape[-1] > target_length:
+                attention_mask = attention_mask[:, :target_length]
```

Comment:
This was needed in case the input prompt is longer than sliding window, so we have to crop attention mask to the max possible length before masking out pad tokens

Reply:
Mmmm I don't think we can crop the attention mask for `eager`

Reply:
For eager we never go beyond cache length because the `target_length` is `attention mask.shape[1]` whenever attention mask is present. So cropping shouldn't yield any difference

Same for static cache, where we expect cache shape to be always more than attention mask

Reply:
ah okay my bad then!